### PR TITLE
feat!: add an implementation of DIP 0027 Credit Asset Locks

### PR DESCRIPTION
**Pull request by knst**
## Issue being fixed or feature implemented
This is an implementation of DIP0027 "Credit Asset Locks".
It's a mechanism to fluidly exchange between Dash and credits.

## What was done?
This pull request includes:
      - Asset Lock transaction
      - Asset Unlock transaction (withdrawal)
      - Credit Pool in coinbase
      - Unit tests for Asset Lock/Unlock tx
      - New functional test `feature_asset_locks.py`

RPC: currently locked amount (credit pool) is available through rpc call `getblock`.

## How Has This Been Tested?
There added new unit tests for basic checks of transaction validity (asset lock/unlock).
Also added new functional test "feature_asset_locks.py" that cover typical cases, but not all corner cases yet.

## Breaking Changes
This feature should be activated as soft-fork because:
 - It adds 2 new special transaction and one of them [asset unlock tx]  requires update consensus rulels
 - It adds new data in coinbase tx (credit pool)

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
**To release DIP 0027**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

**Comment by PastaPastaPasta** [comment]
Please change PR title to follow conventional commits :)

File: src/chainparams.cpp [hunk 1]
```diff
@@ -309,6 +317,8 @@ class CMainParams : public CChainParams {
         consensus.llmqTypeDIP0024InstantSend = Consensus::LLMQType::LLMQ_60_75;
         consensus.llmqTypePlatform = Consensus::LLMQType::LLMQ_100_67;
         consensus.llmqTypeMnhf = Consensus::LLMQType::LLMQ_400_85;
+        // TODO or which one?
+        consensus.llmqTypeAssetLocks = Consensus::LLMQType::LLMQ_400_85;
```
Comment (PastaPastaPasta) [inline src/chainparams.cpp:321]: this is correct
```suggestion
        consensus.llmqTypeAssetLocks = Consensus::LLMQType::LLMQ_400_85;
```

File: src/chainparams.cpp [hunk 2]
```diff
@@ -527,6 +545,8 @@ class CTestNetParams : public CChainParams {
         consensus.llmqTypeDIP0024InstantSend = Consensus::LLMQType::LLMQ_60_75;
         consensus.llmqTypePlatform = Consensus::LLMQType::LLMQ_100_67;
         consensus.llmqTypeMnhf = Consensus::LLMQType::LLMQ_50_60;
+        // TODO or which one?
+        consensus.llmqTypeAssetLocks = Consensus::LLMQType::LLMQ_50_60;
```
Comment (PastaPastaPasta) [inline src/chainparams.cpp:549]: probably fine

File: src/chainparams.cpp [hunk 3]
```diff
@@ -726,6 +754,8 @@ class CDevNetParams : public CChainParams {
...
```
Comment (PastaPastaPasta) [inline src/chainparams.cpp:758]: I'm confused... I thought all of these should be llmq_devnet...

File: src/evo/assetlocktx.cpp [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    CAssetUnlockPayload assetUnlockTx;
+    if (!GetTxPayload(tx, assetUnlockTx)) {
+        return {ValidationInvalidReason::TX_BAD_SPECIAL, "bad-assetunlocktx-payload"};
+    }
+
+    if (assetUnlockTx.getVersion() == 0 || assetUnlockTx.getVersion() > CAssetUnlockPayload::CURRENT_VERSION) {
+        return {ValidationInvalidReason::TX_BAD_SPECIAL, "bad-assetunlocktx-version"};
+    }
+
+    if (assetUnlockTx.getFee() <= 0) {
```
Comment (shumkov) [inline src/evo/assetlocktx.cpp:189]: Should we validate it against dust value?

Comment (thephez) [reply src/evo/assetlocktx.cpp:189]: We should make sure we're not creating dust. This may be checked elsewhere already (e.g. for all transactions), but I'm not sure.

Comment (knst) [reply src/evo/assetlocktx.cpp:189]: this amount is summed up with other fees in block, so, overall fee output from all transaction won't be dust.

Comment (thephez) [reply src/evo/assetlocktx.cpp:189]: Hmm, not sure I understand. It doesn't how many fees are in the block. There shouldn't be any output of any tx in the block below the dust amount. @PastaPastaPasta @UdjinM6 Will miners create blocks that contain dust outputs? Maybe this is a non-issue.

Comment (PastaPastaPasta) [reply src/evo/assetlocktx.cpp:189]: I think what @knst is saying is that the asset lock fee is combined with the block reward, and as such is never dust

**Comment by knst** [comment]
> Please rebase

rebased. Unit tests and functional tests succeed locally.

File: src/bloom.cpp [hunk 5]
```diff
@@ -212,6 +212,10 @@ bool CBloomFilter::CheckSpecialTransactionMatchesAndUpdate(const CTransaction &t
     case(TRANSACTION_MNHF_SIGNAL):
         // No additional checks for this transaction types
         return false;
+    case(TRANSACTION_ASSET_LOCK):
+    case(TRANSACTION_ASSET_UNLOCK):
+        // TODO asset lock/unlock bloom?
```
Comment (PastaPastaPasta) [inline src/bloom.cpp:217]: I think that it's not needed for us to handle bloom for either of these, but @UdjinM6 please provide your thoughts

File: src/evo/assetlocktx.cpp [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    return requestedHeight;
+}
+
+uint256 CAssetUnlockPayload::getQuorumHash() const {
+    return quorumHash;
+}
+
+CBLSSignature CAssetUnlockPayload::getQuorumSig() const {
+    return quorumSig;
+}
```
Comment (PastaPastaPasta) [inline src/evo/assetlocktx.cpp:237]: We probably don't want this, this will copy the object each time

File: src/evo/assetlocktx.h [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+public:
+    CAssetUnlockPayload(uint16_t nVersion, uint64_t index, CAmount fee, uint32_t requestedHeight,
+            uint256 quorumHash, CBLSSignature quorumSig)
+    : nVersion(nVersion),
+    index(index),
+    fee(fee),
+    requestedHeight(requestedHeight),
+    quorumHash(quorumHash),
+    quorumSig(quorumSig)
+    {}
```
Comment (PastaPastaPasta) [inline src/evo/assetlocktx.h:94]: indentation issues, run clang-format probably

File: src/evo/assetlocktx.h [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+    CAssetUnlockPayload() = default;
```
Comment (PastaPastaPasta) [inline src/evo/assetlocktx.h:96]: kinda concerned that quorumHash / quorumSig are uninitialized here

File: src/evo/assetlocktx.h [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+    static constexpr size_t MAXIMUM_WITHDRAWALS = 32;
+
+private:
+    uint16_t nVersion{CURRENT_VERSION};
+    uint64_t index{0};
+    uint32_t fee{0};
+    uint32_t requestedHeight{0};
+    uint256 quorumHash{0};
+    CBLSSignature quorumSig{};
```
Comment (ogabrielides) [inline src/evo/assetlocktx.h:84]: Maybe you can use `CBLSLazySignature` instead for performance. @UdjinM6 what do you think?

**Comment by knst** [comment]
Bumped code to current develop:
 - bumped activation from v19 to v20
 - fixed couple minor bugs and typos
 - squashed several commits/reallocated changes between commits to avoid meaningless changes back-n-forward between commits

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
I'm in the middle of review atm so this is part1

File: src/test/evo_assetlocks_tests.cpp [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+        txManyOutputs.vout.resize(outputsLimit + 1);
+        txManyOutputs.vout.back().nValue = CENT;
+        txManyOutputs.vout.back().scriptPubKey = GetScriptForDestination(PKHash(key.GetPubKey()));
+        BOOST_CHECK(!CheckAssetUnlockTx(CTransaction(txManyOutputs), block_index, pool, tx_state));
+        BOOST_CHECK(tx_state.GetRejectReason() == "bad-assetunlocktx-too-many-outs");
+    }
+
+}
+
+BOOST_FIXTURE_TEST_CASE(evo_skipset, TestChain100Setup)
```
Comment (UdjinM6) [inline src/test/evo_assetlocks_tests.cpp:361]:
```suggestion
BOOST_FIXTURE_TEST_CASE(evo_skipset, BasicTestingSetup)
```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
few more and also pls note that I dropped my previous comments re cbtx version bump

File: src/evo/assetlocktx.cpp [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+    for (const CTxOut& txout : tx.vout) {
+        const CScript& script = txout.scriptPubKey;
+        if (script.empty() || script[0] != OP_RETURN) continue;
+
+        if (script.size() != 2 || script[1] != 0) return state.Invalid(TxValidationResult::TX_BAD_SPECIAL, "bad-assetlocktx-non-empty-return");
+
+        if (txout.nValue <= 0) return state.Invalid(TxValidationResult::TX_BAD_SPECIAL, "bad-assetlocktx-zeroout-return");
+
+        // Should be only one OP_RETURN
+        if (returnAmount) return state.Invalid(TxValidationResult::TX_BAD_SPECIAL, "bad-assetlocktx-multiple-return");
```
[the remark this answers is not in the document]
Comment (knst) [reply src/evo/assetlocktx.cpp:55]: > also, you could probably check this earlier (line 49)

all checks inside this loop are trivial and lightweight: nothing like comparing heavy strings or calculating hashes.

I think it's more organized and logical IMO if comparing `returnAmount` and assigning `returnAmount` will be close to each other.

File: src/evo/creditpool.cpp [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+std::unique_ptr<CCreditPoolManager> creditPoolManager;
+
+static bool getDataFromUnlockTx(const CTransaction& tx, CAmount& toUnlock, uint64_t& index, TxValidationState& state) {
+    CAssetUnlockPayload assetUnlockTx;
+    if (!GetTxPayload(tx, assetUnlockTx)) {
+        return state.Invalid(TxValidationResult::TX_CONSENSUS, "failed-creditpool-unlock-payload");
+    }
+
+    index = assetUnlockTx.getIndex();
+    toUnlock = assetUnlockTx.getFee();
```
[the remark this answers is not in the document]
Comment (knst) [reply src/evo/creditpool.cpp:32]: `getFee()` is uint32, so, it's >= 0 and always in range. DIP says nothing about minimum fee so far as I remember. @shumkov  ?

File: src/evo/assetlocktx.cpp [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+    }
+
+    const auto quorum = llmq::quorumManager->GetQuorum(llmqType, quorumHash);
+    assert(quorum);
+
+    const std::string id(strprintf("plwdtx%lld", index));
+
+    std::vector<uint8_t> vchHash(32);
+    CSHA256().Write(reinterpret_cast<const uint8_t*>(id.data()), id.size()).Finalize(vchHash.data());
+    uint256 requestId(vchHash);
```
[the remark this answers is not in the document]
Comment (knst) [reply src/evo/assetlocktx.cpp:140]: I just didn't know about this feature. Let me try!

File: src/evo/assetlocktx.cpp [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+        return state.Invalid(TxValidationResult::TX_CONSENSUS, "bad-assetunlock-quorum-hash");
+    }
+
+    // Copy transaction except `quorumSig` field to calculate hash
+    CMutableTransaction tx_copy(tx);
+    auto payload_copy = CAssetUnlockPayload(assetUnlockTx.getVersion(), assetUnlockTx.getIndex(), assetUnlockTx.getFee(), assetUnlockTx.getRequestedHeight(), assetUnlockTx.getQuorumHash(), CBLSSignature());
+    SetTxPayload(tx_copy, payload_copy);
+
+    uint256 msgHash = tx_copy.GetHash();
+    return assetUnlockTx.VerifySig(msgHash, pindexPrev, state);
```
[the remark this answers is not in the document]
Comment (knst) [reply src/evo/assetlocktx.cpp:188]: `creditPool` validates that `index` is unique for all mined txes.

File: src/evo/assetlocktx.h [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+    const std::vector<CTxOut>& getCreditOutputs() const;
+};
+
+class CAssetUnlockPayload
+{
+public:
+    static constexpr uint8_t CURRENT_VERSION = 1;
+    static constexpr auto SPECIALTX_TYPE = TRANSACTION_ASSET_UNLOCK;
+
+    static constexpr size_t MAXIMUM_WITHDRAWALS = 32;
```
[the remark this answers is not in the document]
Comment (knst) [reply src/evo/assetlocktx.h:71]: accordingly DIP:

> In order to limit the signing load on validators, there can only be a maximum of 16 withdrawal transactions per platform block. However, with transaction pooling at a maximum of 32 withdrawals per transaction, this can amount to up to 512 withdrawals per block. Assuming a 5 second block this would lead to about 102 withdrawals per second.

File: src/evo/assetlocktx.cpp [hunk 16]
```diff
@@ -115,14 +122,9 @@ bool CAssetUnlockPayload::VerifySig(const uint256& msgHash, const CBlockIndex* p
...
-    if (!llmq_params_opt) {
-        return state.Invalid(TxValidationResult::TX_CONSENSUS, "bad-assetunlock-llmq-type");
-    }
-
-    // We check at most 2 quorums, so, count is equal to 2
-    const int count = 2;
-    auto quorums = llmq::quorumManager->ScanQuorums(llmqType, pindexTip, count > -1 ? count : llmq_params_opt->signingActiveQuorumCount);
+    // We check at most 2 quorums, so, quorum_count is equal to 2
+    const int quorum_count = 2;
+    auto quorums = llmq::quorumManager->ScanQuorums(llmqType, pindexTip, quorum_count);
```
Comment (UdjinM6) [inline src/evo/assetlocktx.cpp:127]: `quorum_count` isn't used anywhere below besides `ScanQuorums`. Why even declaring it? If you want to be clear about the meaning of `2` you can do smth like that instead
```suggestion
    // We check at most 2 quorums
    auto quorums = llmq::quorumManager->ScanQuorums(llmqType, pindexTip, 2 /* nCountRequested */);
```

**Review by UdjinM6 — COMMENTED** [review]
Looks good overall but I want to re-review it (CreditPool logic specifically) once again with fresh eyes tomorrow.

**Comment by UdjinM6** [comment]
`Check Merge Fast-Forward Only` failed again 👀

File: src/evo/assetlocktx.cpp (hunk 16 above)
Comment (knst) [reply src/evo/assetlocktx.cpp:127]: I just don't want really hanging literal that even doesn't have a type, but it doesn't really matter/useful.

File: src/evo/assetlocktx.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (UdjinM6) [reply src/evo/assetlocktx.cpp]: Actually... I misunderstood what's going on here initially so my suggestion is wrong (and it broke tests). Let's revert it 3bea3e1def.

File: src/evo/creditpool.cpp [hunk 17]
```diff
@@ -0,0 +1,344 @@
...
+    }
+    current_max = value + 1;
+    return true;
+}
+
+bool CSkipSet::canBeAdded(uint64_t value) const
+{
+    if (contains(value)) return false;
+
+    if (skipped.find(value) != skipped.end()) return true;
```
[the remark this answers is not in the document]
Comment (knst) [reply src/evo/creditpool.cpp:97]: not exactly. If we will add element in the middle (when this value is already skipped!) in this case no extra memory will be used, more over memory usage will be reduced. So, even if there's no any extra space, the value may be added.

On other hand, in this condition:
`if (capacity() + value - current_max > capacity_limit) {`
Let's assume that capacity is 1, value is 1, current_max is 20.
data structure have all numbers from 2 to 20. Internal presentation is: current_max = 20, skipped: {1}.
Let's try to add value "1": `1 + 1 - 20 = negative number`  and comparision will fail (because all numbers here unsigned).

So, value can't be added.
-------------------------
@UdjinM6 
Maybe replace here "skipping list" structure to "ranges"? such as [2, 20), [21, 45), [50, 51). It's easy to provide auto-merging ranges when elements are added; performance is similar.
**Pros:** Adding operations should not care if there's somehow `index` with big gap - it won't eat ram accidentally
**Cons:** eats more memories in case if gaps are small (1 number only) - consume memory for 2 number (left/right)  instead 1 number.

also it won't stuck all asset-unlocks txes once if capacity will run out - it will just eat little more memory (linear from amount of gaps, but independent how big they are).

File: src/evo/creditpool.cpp [hunk 18]
```diff
@@ -0,0 +1,344 @@
...
+    if (sessionUnlocked + toUnlock > pool.currentLimit) {
+        return state.Invalid(TxValidationResult::TX_CONSENSUS, "failed-creditpool-unlock-too-much");
+    }
+
+    if (pool.indexes.contains(index) || newIndexes.count(index)) {
+        return state.Invalid(TxValidationResult::TX_CONSENSUS, "failed-creditpool-duplicated-index");
+    }
+
+    if (!pool.indexes.canBeAdded(index)) {
+        return state.Invalid(TxValidationResult::TX_CONSENSUS, "failed-getcbforblock-index-exceed");
```
[the remark this answers is not in the document]
Comment (knst) [reply src/evo/creditpool.cpp:311]: `getcbforblock` some time ago it was a function in miner.cpp; refactored later to creditpool. Sorry for confusion

File: src/evo/creditpool.h [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+    static constexpr int LimitBlocksToTrace = 576;
+    static constexpr CAmount LimitAmountLow = 100 * COIN;
+    static constexpr CAmount LimitAmountHigh = 1000 * COIN;
+
+    explicit CCreditPoolManager(CEvoDB& _evoDb);
+
+    ~CCreditPoolManager() = default;
+
+    /**
+      * @return CCreditPool with data or with empty depends on activation V19 at that block
```
[the remark this answers is not in the document]
Comment (knst) [reply src/evo/creditpool.h:160]: firstly, I just want to distinct for sure `not available` and `empty` values; because there's cache of credit pools in `creditPoolCache` and empty nullopt means that object is not found in cache. I feel like it is more error prune.

Secondly, even more important, since v20 activated - the value in CbTx should be always presented (even zero); before v20 - should not be even calculated.

So, empty and `nullopt` have different semantic meaning.

**Review by UdjinM6 — DISMISSED** [review]
LGTM, utACK

File: src/evo/assetlocktx.cpp [hunk 20]
```diff
@@ -0,0 +1,249 @@
...
+    if (tx.vout.size() > CAssetUnlockPayload::MAXIMUM_WITHDRAWALS) {
+        return state.Invalid(TxValidationResult::TX_BAD_SPECIAL, "bad-assetunlocktx-too-many-outs");
+    }
+
+    CAssetUnlockPayload assetUnlockTx;
+    if (!GetTxPayload(tx, assetUnlockTx)) {
+        return state.Invalid(TxValidationResult::TX_BAD_SPECIAL, "bad-assetunlocktx-payload");
+    }
+
+    if (assetUnlockTx.getVersion() == 0 || assetUnlockTx.getVersion() > CAssetUnlockPayload::CURRENT_VERSION) {
```
Comment (PastaPastaPasta) [inline src/evo/assetlocktx.cpp:173]: please don't use public getters like this when we have access to the actual variable; this isn't java :D I'd prefer to have minimal getter usage

Comment (PastaPastaPasta) [reply src/evo/assetlocktx.cpp:173]: Okay, nevermind; I see we don't have access here... I don't love the getter usage but I guess that's fine

File: src/evo/assetlocktx.cpp [hunk 21]
```diff
@@ -0,0 +1,249 @@
...
+
+const CBLSSignature& CAssetUnlockPayload::getQuorumSig() const
+{
+    return quorumSig;
+}
+
+int CAssetUnlockPayload::getHeightToExpiry() const
+{
+    return requestedHeight + HEIGHT_DIFF_EXPIRING;
+}
```
Comment (PastaPastaPasta) [inline src/evo/assetlocktx.cpp:249]: let's define this in the header as they're trivial methods

File: src/evo/assetlocktx.cpp (hunk 20 above)
Comment (knst) [reply src/evo/assetlocktx.cpp:173]: I can't make members constant, because READWRITE macros for serialization/de-serialization.

Because the members of `CAsset{Lock,Unlock}Payload` are not constant, better to have them private -> need to have getters. If there were other way to do serialization/deserialization without bunch of boiler code I'd like to get rid of these  getters too

File: src/evo/assetlocktx.cpp [hunk 22]
```diff
@@ -0,0 +1,205 @@
...
+#include <llmq/commitment.h>
+#include <llmq/signing.h>
+#include <llmq/utils.h>
+#include <llmq/quorums.h>
+
+#include <algorithm>
+
+/*
+   Common code for Asset Lock and Asset Unlock
+    */
```
Comment (ogabrielides) [inline src/evo/assetlocktx.cpp:24]: end of comment

File: src/evo/assetlocktx.cpp [hunk 23]
```diff
@@ -0,0 +1,205 @@
...
+    case TRANSACTION_ASSET_UNLOCK:
+        return CheckAssetUnlockTx(tx, pindexPrev, creditPool, state);
+    default:
+        return state.Invalid(TxValidationResult::TX_BAD_SPECIAL, "bad-not-asset-locks-at-all");
+    }
+}
+
+/*
+   Asset Lock Transaction
+   */
```
Comment (ogabrielides) [inline src/evo/assetlocktx.cpp:39]: end of comment

File: src/evo/assetlocktx.cpp [hunk 24]
```diff
@@ -0,0 +1,205 @@
...
+        outputs.append(",");
+    }
+    outputs.back() = ']';
+    return strprintf("CAssetLockPayload(nVersion=%d,creditOutputs=%s)", nVersion, outputs.c_str());
+}
+
+/*
+   Asset Unlock Transaction (withdrawals)
+   */
+
```
Comment (ogabrielides) [inline src/evo/assetlocktx.cpp:107]: end of comment + delete empty line

File: src/evo/creditpool.cpp [hunk 25]
```diff
@@ -0,0 +1,302 @@
...
+    }
+}
+
+static std::optional<CBlock> GetBlockForCreditPool(const CBlockIndex* const block_index, const Consensus::Params& consensusParams)
+{
+    CBlock block;
+    if (!ReadBlockFromDisk(block, block_index, consensusParams)) {
+        throw std::runtime_error("failed-getcbforblock-read");
+    }
+    // Should not fail if V20 (DIP0027) are active but it happens for Unit Tests
```
Comment (ogabrielides) [inline src/evo/creditpool.cpp:115]: is

File: src/evo/creditpool.cpp [hunk 26]
```diff
@@ -0,0 +1,302 @@
...
+    assert(!block.vtx[0]->vExtraPayload.empty());
+
+    return block;
+}
+
+CCreditPool CCreditPoolManager::ConstructCreditPool(const CBlockIndex* const block_index, CCreditPool prev, const Consensus::Params& consensusParams)
+{
+    std::optional<CBlock> block = GetBlockForCreditPool(block_index, consensusParams);
+    if (!block) {
+        // If reading of previous block is not read successfully, but
```
Comment (ogabrielides) [inline src/evo/creditpool.cpp:129]: is not successful

**Review by ogabrielides — COMMENTED** [review]
LGTM in general. A few very minor comments

File: src/evo/specialtxman.cpp [hunk 27]
```diff
@@ -127,6 +149,23 @@ bool ProcessSpecialTxsInBlock(const CBlock& block, const CBlockIndex* pindex, ll
                 return state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, tx_state.GetRejectReason(),
                                  strprintf("Process Special Transaction failed (tx hash %s) %s", ptr_tx->GetHash().ToString(), tx_state.GetDebugMessage()));
             }
+            if (creditPoolDiff != std::nullopt && !creditPoolDiff->ProcessTransaction(*ptr_tx, tx_state)) {
+                assert(tx_state.GetResult() == TxValidationResult::TX_CONSENSUS || tx_state.GetResult() == TxValidationResult::TX_BAD_SPECIAL);
+                return state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, tx_state.GetRejectReason(),
+                                 strprintf("Process Special Transaction failed at Credit Pool (tx hash %s) %s", ptr_tx->GetHash().ToString(), tx_state.GetDebugMessage()));
+
+                return false;
```
Comment (UdjinM6) [inline src/evo/specialtxman.cpp:157]:
```suggestion
```

File: src/evo/creditpool.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (UdjinM6) [reply src/evo/creditpool.cpp:33]: ping re min fee

[the remark this answers is not in the document]
Comment (knst) [reply src/evo/creditpool.cpp:33]: as I talked with @shumkov: if an asset-unlock transaction will not be mined by any reason and timeout will happen - this transaction is changing state to "EXPIRY" and it is a terminal state for transaction.
In this case platform may generate new transaction for taking away credits from platform to blockchain (by new asset unlock transaction). New transaction can have higher fee to make miners be more interested in including this transaction in block.

Asset-unlock transaction has no inputs, but have outputs + spare field "fee" (getFee).
1. Good to check that all outputs are above dust value - because it's just regular output of tx.
2. this fee is not used separately, but sum-up will values of all other transaction and are used as part of "block reward". Even if transaction will have a fee 0.00000001 it won't be dust, because this fee will sum up with other values. 
3. there's validation of fee: `Consensus::CheckTxInputs` - this function setup value of `txfee` correctly and this fee checked again mempool limits, etc.

Beside that,  if any miner will include transaction with fee below `minRelayFee` - it doesn't mean that transaction is invalid. It is just miner decided to take tx with low fee, not validation's business - tx is valid.

[the remark this answers is not in the document]
Comment (UdjinM6) [reply src/evo/creditpool.cpp:33]: Thinking more about fees and stuff… we should change `feature_asset_locks.py` to avoid mining and creating/sending txes on the same node i.e. it should be 5 nodes total: 1. controller/mining node 2. wallet node 3,4,5. MNs. Currently we create and mine txes on the very same node (`self.nodes[0]`) so we don’t really see if smth stops these txes from being relayed or not. Lock txes should be created on the wallet node and unlock txes should be created on one of MNs. Also would be nice to see a test for unlock txes with different `fee`s (from 0 to >0.1 DASH).

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
https://github.com/dashpay/dash/pull/5026/files#r1267324806

**Comment by UdjinM6** [comment]
Looks good! 👍 

What about
> Also would be nice to see a test for unlock txes with different fees (from 0 to >0.1 DASH).

?

**Comment by knst** [comment]
> Also would be nice to see a test for unlock txes with different fees (from 0 to >0.1 DASH).

added just now, thanks for reminding!

**Comment by UdjinM6** [comment]
Ok, so this
```
diff --git a/test/functional/feature_asset_locks.py b/test/functional/feature_asset_locks.py
index 725d290f6e..32176e309c 100755
--- a/test/functional/feature_asset_locks.py
+++ b/test/functional/feature_asset_locks.py
@@ -295,7 +295,7 @@ class AssetLocksTest(DashTestFramework):
 
         self.log.info("Generating several txes by same quorum....")
         self.validate_credit_pool_amount(locked_1)
-        asset_unlock_tx = self.create_assetunlock(101, COIN, pubkey)
+        asset_unlock_tx = self.create_assetunlock(101, COIN, pubkey, fee=1)
         asset_unlock_tx_late = self.create_assetunlock(102, COIN, pubkey)
         asset_unlock_tx_too_late = self.create_assetunlock(103, COIN, pubkey)
         asset_unlock_tx_too_big_fee = self.create_assetunlock(104, COIN, pubkey, fee=int(Decimal("0.1") * COIN))
```
results in failures like
> AssertionError: not([{'allowed': True, 'txid': '27d3357231f0f877d69a94d72d5f0e374a40fe8936974f868ba069439eb13aea'}] == [{'txid': '27d3357231f0f877d69a94d72d5f0e374a40fe8936974f868ba069439eb13aea', 'allowed': False, 'reject-reason': 'min relay fee not met'}])

That's exactly what I was worried about. 0f35ff36e3 should help. Or maybe we need to think more about unlocking vs min fees.

File: src/evo/creditpool.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (knst) [reply src/evo/creditpool.cpp:33]: Why do we need to bypass limit? If fee is too low and tx is not mined - platform will generate new one after expiration. There's nothing special about asset-unlock.
 Tests works as I expected when implemented it...

For some txes such as quorum/etc there are nowhere to take reward/fee, so, checks are bypassed. But unlock txes are not special in this terms.
If fee is too small - it's okay to request "more fee, please"

**Comment by UdjinM6** [comment]
> Why do we need to bypass limit? If fee is too low and tx is not mined - platform will generate new one after expiration. There's nothing special about asset-unlock. Tests works as I expected when implemented it...
> 
> For some txes such as quorum/etc there are nowhere to take reward/fee, so, checks are bypassed. But unlock txes are not special in this terms. If fee is too small - it's okay to request "more fee, please"

https://github.com/dashpay/dash/pull/5026/files#r1269915571

oh, so it's ok for unlock tx to fail? even when fees are up or mempool is full?

**Review by UdjinM6 — APPROVED** [review]
utACK then :)

**Review by ogabrielides — APPROVED** [review]
re-utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for squash merge

File: src/evo/cbtx.h [hunk 28]
```diff
@@ -62,6 +68,7 @@ class CCbTx
             if (nVersion >= CB_V20_VERSION) {
                 obj.pushKV("bestCLHeightDiff", static_cast<int>(bestCLHeightDiff));
                 obj.pushKV("bestCLSignature", bestCLSignature.ToString());
+                obj.pushKV("assetLockedAmount", ValueFromAmount(assetLockedAmount));
```
Comment (thephez) [inline src/evo/cbtx.h:71]: @knst @PastaPastaPasta @ogabrielides @UdjinM6 A little late to the party here, but what do you think about calling it smth like `creditPoolBalance` instead? It wasn't clear to me at first if "assetLockedAmount" referred just to the asset locks in this block or the system-wide asset lock balance.

Comment (knst) [reply src/evo/cbtx.h:71]: I agree about `creditPoolBalance` - maybe that's not perfect name still, but definitely better than assetLockedAmount.

Let's gather feedback from everyone and I will prepare PR to change it.

Comment (ogabrielides) [reply src/evo/cbtx.h:71]: Yup, I like `creditPoolBalance `

Comment (UdjinM6) [reply src/evo/cbtx.h:71]: makes sense

Comment (knst) [reply src/evo/cbtx.h:71]: Here it does: https://github.com/dashpay/dash/pull/5526
